### ISSUE TITLE
`nufft2` gradient is incorrect under `shard_map` (forward is correct)

### ISSUE THREAD

**Issue by rohandahale** [flatironinstitute/jax-finufft#226]
## TL;DR

Inside a `jax.shard_map` (a.k.a. `jax.experimental.shard_map.shard_map`), the **forward** `nufft2` is bit-exact vs single-device, but the **gradient** (via `jax.grad` / `jax.value_and_grad`) is wrong. The same computation single-device is correct. This makes `nufft2` unusable for multi-GPU data-parallel objectives that differentiate through it under `shard_map`.

## Environment

- `jax` 0.10.1, `jaxlib` 0.10.1
- `jax-finufft` 1.3.1, `finufft` 2.5.1
- 4× NVIDIA GPU (CUDA), `jax.config.update("jax_enable_x64", True)`

## Minimal reproduction

Data-parallel over the nonuniform points: the image grid is **replicated**, the points `x`/`y` and the data are **sharded** across the mesh; each device runs a local `nufft2` and the loss is reduced with `psum`. Needs ≥2 devices (real GPUs, or simulate on CPU with `XLA_FLAGS=--xla_force_host_platform_device_count=2`).

```python
import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp
import numpy as np
from jax.sharding import NamedSharding, PartitionSpec as P
from jax import shard_map  # or: from jax.experimental.shard_map import shard_map
from jax_finufft import nufft2

mesh = jax.sharding.Mesh(np.array(jax.devices()), ("s",))
k = mesh.size
N, xd, yd, eps = k * 257, 16, 16, 1e-12
rng = np.random.default_rng(0)
uv = (rng.random((N, 2)) - 0.5) * 2 * np.pi
grid = jnp.asarray(rng.standard_normal((xd, yd)) + 1j * rng.standard_normal((xd, yd)))
ux, uy = jnp.asarray(uv[:, 0]), jnp.asarray(uv[:, 1])
data = jnp.asarray(rng.standard_normal(N) + 1j * rng.standard_normal(N))

def chi(g, x, y, d):
    return jnp.sum(jnp.abs(nufft2(g, x, y, iflag=-1, eps=eps) - d) ** 2)

# reference: single device
v0, g0 = jax.value_and_grad(lambda g: chi(g, ux, uy, data))(grid)

# sharded over the points, image replicated, loss reduced with psum
rows = NamedSharding(mesh, P("s"))
uxs, uys, ds = (jax.device_put(ux, rows), jax.device_put(uy, rows), jax.device_put(data, rows))
gridrep = jax.device_put(grid, NamedSharding(mesh, P()))

def local(g, x, y, d):
    return jax.lax.psum(chi(g, x, y, d), "s")

sm = shard_map(local, mesh=mesh, in_specs=(P(), P("s"), P("s"), P("s")), out_specs=P())
v1, g1 = jax.value_and_grad(lambda g: sm(g, uxs, uys, ds))(gridrep)

print("value  rel = %.2e" % (abs(float(v1) - float(v0)) / abs(float(v0))))   # ~1e-16  (forward OK)
print("grad   rel = %.2e" % float(jnp.linalg.norm(g1 - g0) / jnp.linalg.norm(g0)))  # ~9e-1  (gradient WRONG)
```

### Observed output (4-device mesh)

```
value  rel = 1.2e-16     # forward bit-exact
grad   rel = 8.7e-01     # gradient ~86% off
```

## Diagnosis

The `nufft2` forward custom call shards correctly (each device evaluates a local type-2 transform on its slice of points — the value is exact). The problem is the **registered transpose** of the `nufft2` primitive: under `shard_map`'s SPMD autodiff it mis-accumulates the cotangent of the replicated `source` (the grid). We confirmed the error is independent of how the gradient is taken — outer transpose through the `shard_map`, an inner `jax.value_and_grad` inside the `shard_map` body, and a `custom_vjp` wrapping `nufft2` all give the same wrong result. A dense `jnp.dot`-based equivalent in the identical harness is bit-exact, which points at the custom-call transpose rather than the harness.

## Workaround (for reference)

Bypassing the transpose entirely is correct to machine precision: compute the sample-space cotangent `c = ∂loss/∂samples` with ordinary ops (no nufft), then apply the adjoint as an **explicit forward** `nufft1` and `psum` it. Since type-1 is a sum over points, it is additive across shards, so `psum(nufft1(c_local, x_local, y_local)) == nufft1(c_full, x_full, y_full)`:

```python
def grad_local(g, x, y, d):
    s = nufft2(g, x, y, iflag=-1, eps=eps)
    c = 2 * jnp.conj(s - d)                      # ∂ sum|s-d|^2 / ∂s
    return jax.lax.psum(nufft1(grid.shape, c, x, y, iflag=-1, eps=eps), "s")
# shard_map(grad_local, ...) matches the single-device gradient to ~5e-16
```

This works, but it would be much cleaner if the registered `nufft2`/`nufft1` transposes were `shard_map`-correct so that plain `jax.grad` worked under SPMD.

## Ask

Should `nufft1`/`nufft2`/`nufft3` transposes be SPMD-aware so they compose with `shard_map`, or is data-parallel autodiff over the points out of scope (and worth a docs note + the forward-adjoint workaround)?

**Comment by lgarrison** [comment]
Thanks for the report! I think this is related to #179, i.e. the behavior of FFI under sharding. The silently wrong results are obviously concerning, though, and we should see if we can do better (or at least document it).

The solution Claude prefers is essentially:
```python
# ops.py
from jax._src.core import auto_insert_reshard

def nufft1(output_shape, source, *points, iflag=1, eps=1e-6, opts=None):
    ...
    # Execute the transform primitive
    source, *points = auto_insert_reshard(source, *points)
    result = nufft1_p.bind(source, *points, ...)
    ...
```
And likewise for nufft2 and nufft3.

However, this is a bleeding-edge area in which Claude is not very confident in its own abilities (and neither am I)! The `auto_insert_reshard` function was only introduced in the last release, 0.10.1, less than 1 month ago (renamed from `standard_insert_pvary`, introduced in 0.10.0).

So, before we go any further, perhaps we should wait a bit to see if @dfm or any other JAX FFI experts would like to chime in on the right approach?

**Comment by dfm** [comment]
Yeah I think something like that is what I would try! I'm not totally up to date on exactly which functions to use, but this seems like the right direction.
